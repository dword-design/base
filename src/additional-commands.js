import { mapValues } from '@dword-design/functions'

import commit from './commit'
import config from './config'
import lint from './lint'
import test from './test'

export default {
  commit: {
    handler: commit,
    options: [{ description: 'Allow empty commits', name: '--allow-empty' }],
  },
  lint: {
    handler: lint,
  },
  test: {
    arguments: '[pattern]',
    handler: test,
    options: [
      {
        description: 'Only run tests matching this string or regexp',
        name: '-g, --grep <grep>',
      },
    ],
  },
  ...(config.commands
    |> mapValues(command =>
      typeof command === 'function' ? { handler: command } : command
    )),
}
