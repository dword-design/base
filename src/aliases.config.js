module.exports = {
  '@variables': require('./variables').path,
  '@functions': require.resolve('@dword-design/functions'),
}
