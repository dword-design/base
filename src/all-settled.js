const allSettled = require('promise.allsettled')
const { find } = require('lodash')

module.exports = promises => allSettled(promises)
  .then(results => {
    const rejected = find(results, { status: 'rejected' })
    if (rejected !== undefined) {
      throw rejected.reason
    }
  })
