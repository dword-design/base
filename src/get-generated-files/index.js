import { join, jsonToString, map } from '@dword-design/functions'
import sortKeys from 'sort-keys'
import sortPackageJson from 'sort-package-json'
import yaml from 'yaml'

import babelConfig from './babel.js'
import commitizenConfig from './commitizen.js'
import commitlintConfig from './commitlint.js'
import editorconfigConfig from './editorconfig.js'
import gitattributesConfig from './gitattributes.js'
import githubDeprecatedDependenciesConfig from './github-deprecated-dependencies/index.js'
import deprecatedDependenciesIssueTemplate from './github-deprecated-dependencies-issue-template.js'
import githubFunding from './github-funding.js'
import githubLabelsConfig from './github-labels.js'
import githubSyncLabelsConfig from './github-sync-labels.js'
import gitpodConfig from './gitpod.js'

export default function () {
  const packageConfig = this.getPackageConfig()

  return {
    '.babelrc.json': `${babelConfig |> jsonToString({ indent: 2 })}\n`,
    '.commitlintrc.json': `${
      commitlintConfig |> jsonToString({ indent: 2 })
    }\n`,
    '.cz.json': `${commitizenConfig |> jsonToString({ indent: 2 })}\n`,
    '.devcontainer/devcontainer.json': `${
      this.getGithubCodespacesConfig() |> jsonToString({ indent: 2 })
    }\n`,
    '.editorconfig': editorconfigConfig,
    '.eslintrc.json': `${
      this.getEslintConfig() |> jsonToString({ indent: 2 })
    }\n`,
    '.gitattributes': gitattributesConfig,
    '.github/DEPRECATED_DEPENDENCIES_ISSUE_TEMPLATE.md':
      deprecatedDependenciesIssueTemplate,
    '.github/FUNDING.yml': githubFunding |> yaml.stringify,
    '.github/labels.yml':
      sortKeys(githubLabelsConfig, { deep: true }) |> yaml.stringify,
    '.github/workflows/build.yml':
      sortKeys(this.getGithubWorkflowConfig(), { deep: true })
      |> yaml.stringify,
    '.github/workflows/deprecated-dependencies.yml':
      sortKeys(githubDeprecatedDependenciesConfig, { deep: true })
      |> yaml.stringify,
    '.github/workflows/sync-labels.yml':
      sortKeys(githubSyncLabelsConfig, { deep: true }) |> yaml.stringify,
    '.github/workflows/sync-metadata.yml':
      sortKeys(this.getGithubSyncMetadataConfig(), { deep: true })
      |> yaml.stringify,
    '.gitignore':
      this.getGitignoreConfig() |> map(entry => `${entry}\n`) |> join(''),
    '.gitpod.Dockerfile': this.getGitpodDockerfile(),
    '.gitpod.yml': gitpodConfig |> yaml.stringify,
    '.releaserc.json': `${
      this.getReleaseConfig() |> jsonToString({ indent: 2 })
    }\n`,
    '.renovaterc.json': `${
      sortKeys(this.getRenovateConfig(), { deep: true })
      |> jsonToString({ indent: 2 })
    }\n`,
    '.vscode/settings.json': `${
      this.getVscodeConfig() |> jsonToString({ indent: 2 })
    }\n`,
    'LICENSE.md': this.getLicenseString(),
    'README.md': this.getReadmeString(),
    'package.json': `${
      packageConfig |> sortPackageJson |> jsonToString({ indent: 2 })
    }\n`,
  }
}
