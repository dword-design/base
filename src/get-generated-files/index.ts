import sortKeys from 'sort-keys';
import sortPackageJson from 'sort-package-json';
import { stringify as yamlStringify } from 'yaml';

import commitizenConfig from './commitizen';
import commitlintConfig from './commitlint';
import editorconfigConfig from './editorconfig';
import gitattributesConfig from './gitattributes';
import githubDeprecatedDependenciesConfig from './github-deprecated-dependencies';
import deprecatedDependenciesIssueTemplate from './github-deprecated-dependencies-issue-template';
import githubFunding from './github-funding';
import githubLabelsConfig from './github-labels';
import githubSyncLabelsConfig from './github-sync-labels';
import npmrc from './npmrc';

export default function () {
  const packageConfig = this.getPackageConfig();
  return {
    '.commitlintrc.json': `${JSON.stringify(commitlintConfig, undefined, 2)}\n`,
    '.cz.json': `${JSON.stringify(commitizenConfig, undefined, 2)}\n`,
    '.devcontainer/devcontainer.json': `${JSON.stringify(
      this.githubCodespacesConfig,
      undefined,
      2,
    )}\n`,
    '.editorconfig': editorconfigConfig,
    '.gitattributes': gitattributesConfig,
    '.github/DEPRECATED_DEPENDENCIES_ISSUE_TEMPLATE.md':
      deprecatedDependenciesIssueTemplate,
    '.github/FUNDING.yml': yamlStringify(githubFunding),
    '.github/labels.yml': yamlStringify(
      sortKeys(githubLabelsConfig, { deep: true }),
    ),
    '.github/workflows/build.yml': yamlStringify(
      sortKeys(this.getGithubWorkflowConfig(), { deep: true }),
    ),
    '.github/workflows/deprecated-dependencies.yml': yamlStringify(
      sortKeys(githubDeprecatedDependenciesConfig, { deep: true }),
    ),
    '.github/workflows/sync-labels.yml': yamlStringify(
      sortKeys(githubSyncLabelsConfig, { deep: true }),
    ),
    '.github/workflows/sync-metadata.yml': yamlStringify(
      sortKeys(this.getGithubSyncMetadataConfig(), { deep: true }),
    ),
    '.gitignore': this.getGitignoreConfig()
      .map(entry => `${entry}\n`)
      .join(''),
    '.gitpod.Dockerfile': this.getGitpodDockerfile(),
    '.gitpod.yml': yamlStringify(this.getGitpodConfig()),
    '.npmrc': `${npmrc}\n`,
    '.releaserc.json': `${JSON.stringify(
      this.getReleaseConfig(),
      undefined,
      2,
    )}\n`,
    '.renovaterc.json': `${JSON.stringify(
      sortKeys(this.getRenovateConfig(), { deep: true }),
      undefined,
      2,
    )}\n`,
    '.vscode/settings.json': `${JSON.stringify(
      this.getVscodeConfig(),
      undefined,
      2,
    )}\n`,
    'LICENSE.md': this.getLicenseString(),
    'README.md': this.getReadmeString(),
    'eslint.config.ts': this.getEslintConfig(),
    'package.json': `${JSON.stringify(
      sortPackageJson(packageConfig),
      undefined,
      2,
    )}\n`,
    'tsconfig.json': `${JSON.stringify(sortKeys(this.getTypescriptConfig(), { deep: true }), undefined, 2)}\n`,
  };
}
