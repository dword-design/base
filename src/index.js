import babelConfig from '@dword-design/babel-config'
import { endent, identity } from '@dword-design/functions'
import deepmerge from 'deepmerge'
import depcheck from 'depcheck'
import depcheckDetectorExeca from 'depcheck-detector-execa'
import depcheckDetectorPackageName from 'depcheck-detector-package-name'
import packageName from 'depcheck-package-name'
import depcheckParserBabel from 'depcheck-parser-babel'
import jiti from 'jiti'
import loadPkg from 'load-pkg'
import P from 'path'
import { transform as pluginNameToPackageName } from 'plugin-name-to-package-name'

import checkUnknownFiles from './commands/check-unknown-files/index.js'
import commit from './commands/commit/index.js'
import depcheckMethod from './commands/depcheck/index.js'
import lint from './commands/lint/index.js'
import prepare from './commands/prepare/index.js'
import test from './commands/test/index.js'
import testDocker from './commands/test-docker/index.js'
import testRaw from './commands/test-raw/index.js'
import getDepcheckSpecialBase from './get-depcheck-special-base/index.js'
import getEditorIgnoreConfig from './get-generated-files/get-editor-ignore/index.js'
import getEslintConfig from './get-generated-files/get-eslint/index.js'
import getGithubCodespacesDockerfile from './get-generated-files/get-github-codespaces-dockerfile/index.js'
import getGithubSyncMetadataConfig from './get-generated-files/get-github-sync-metadata/index.js'
import getGithubWorkflowConfig from './get-generated-files/get-github-workflow/index.js'
import getGitignoreConfig from './get-generated-files/get-gitignore/index.js'
import getGitpodDockerfile from './get-generated-files/get-gitpod-dockerfile.js'
import getLicenseString from './get-generated-files/get-license-string.js'
import getPackageConfig from './get-generated-files/get-package-config/index.js'
import getReadmeString from './get-generated-files/get-readme-string/index.js'
import getReleaseConfig from './get-generated-files/get-release/index.js'
import getRenovateConfig from './get-generated-files/get-renovate/index.js'
import getVscodeConfig from './get-generated-files/get-vscode/index.js'
import getGeneratedFiles from './get-generated-files/index.js'
import getGitInfo from './get-git-info/index.js'

class Base {
  constructor(config) {
    const jitiInstance = jiti(process.cwd(), {
      esmResolve: true,
      interopDefault: true,
      transformOptions: {
        babel: babelConfig,
      },
    })
    if (config === undefined) {
      config = { name: packageName`@dword-design/base-config-node` }
    }
    if (typeof config === 'function') {
      config = config()
    }
    if (config.name) {
      config.name = pluginNameToPackageName(config.name, 'base-config')
    }
    this.packageConfig = loadPkg.sync() || {}

    const defaultConfig = {
      allowedMatches: [],
      commands: {},
      coverageFileExtensions: [],
      depcheckConfig: {
        detectors: [
          depcheck.detector.importDeclaration,
          depcheck.detector.requireCallExpression,
          depcheck.detector.requireResolveCallExpression,
          depcheckDetectorExeca,
          depcheckDetectorPackageName,
        ],
        ignorePath: '.gitignore',
        parsers: {
          '**/*.js': depcheckParserBabel,
        },
        specials: [getDepcheckSpecialBase(config.name), depcheck.special.bin],
      },
      deployAssets: [],
      deployEnv: {},
      deployPlugins: [],
      editorIgnore: [],
      git: getGitInfo(),
      gitignore: [],
      lint: identity,
      nodeVersion: 16,
      preDeploySteps: [],
      prepare: identity,
      readmeInstallString: endent`
        ## Install

        \`\`\`bash
        # npm
        $ npm install ${config.global ? '-g ' : ''}${this.packageConfig.name}

        # Yarn
        $ yarn ${config.global ? 'global ' : ''}add ${this.packageConfig.name}
        \`\`\`
      `,
      seeAlso: [],
      supportedNodeVersions: [14, 16, 18],
      syncKeywords: true,
      windows: true,
    }

    const mergeOptions = {
      customMerge: key =>
        key === 'supportedNodeVersions' ? (a, b) => b : undefined,
    }

    const configsToMerge = [defaultConfig]
    if (config.name) {
      const inheritedConfigPath =
        config.name === this.packageConfig.name
          ? P.resolve('src', 'index.js')
          : config.name
      let inheritedConfig = inheritedConfigPath
        ? jitiInstance(inheritedConfigPath)
        : undefined
      if (typeof inheritedConfig === 'function') {
        inheritedConfig = inheritedConfig(
          deepmerge(defaultConfig, config, mergeOptions)
        )
      }
      configsToMerge.push(inheritedConfig)
    }
    configsToMerge.push(config)
    this.config = deepmerge.all(configsToMerge, mergeOptions)
    this.packageConfig = this.getPackageConfig()
    this.generatedFiles = this.getGeneratedFiles()
  }
}
Object.assign(Base.prototype, {
  checkUnknownFiles,
  commit,
  depcheck: depcheckMethod,
  getEditorIgnoreConfig,
  getEslintConfig,
  getGeneratedFiles,
  getGithubCodespacesDockerfile,
  getGithubSyncMetadataConfig,
  getGithubWorkflowConfig,
  getGitignoreConfig,
  getGitpodDockerfile,
  getLicenseString,
  getPackageConfig,
  getReadmeString,
  getReleaseConfig,
  getRenovateConfig,
  getVscodeConfig,
  lint,
  prepare,
  test,
  testDocker,
  testRaw,
})

export { Base }
