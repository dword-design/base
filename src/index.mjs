import { mapValues } from '@dword-design/functions'

import commands from './commands/index.mjs'

export default commands |> mapValues('handler')
