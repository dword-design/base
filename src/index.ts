import pathLib from 'node:path';

import { createDefu } from '@dword-design/defu';
import type { Options as DepcheckOptions } from 'depcheck';
import depcheck from 'depcheck';
import depcheckDetectorBinName from 'depcheck-detector-bin-name';
import depcheckDetectorExeca from 'depcheck-detector-execa';
import depcheckDetectorPackageName from 'depcheck-detector-package-name';
import packageName from 'depcheck-package-name';
import endent from 'endent';
import fs from 'fs-extra';
import type GitHost from 'hosted-git-info';
import { createJiti } from 'jiti';
import type { Configuration as LintStagedConfig } from 'lint-staged';
import { identity, mapValues } from 'lodash-es';
import type { PartialCommandObjectInObject } from 'make-cli';
import { transform as pluginNameToPackageName } from 'plugin-name-to-package-name';
import type { RenovateConfig } from 'renovate/dist/config/types';
import type { PackageJson, TsConfigJson } from 'type-fest';

import checkUnknownFiles from './commands/check-unknown-files';
import type { PartialCommandOptions } from './commands/partial-command-options';
import commit from './commands/commit';
import depcheckMethod from './commands/depcheck';
import lint from './commands/lint';
import lintPackagejson from './commands/lint/lint-packagejson';
import prepare from './commands/prepare';
import test from './commands/test';
import testDocker from './commands/test-docker';
import testRaw from './commands/test-raw';
import typecheck from './commands/typecheck';
import verify from './commands/verify';
import getDepcheckSpecialBase from './get-depcheck-special-base';
import getGeneratedFiles from './get-generated-files';
import getEditorIgnoreConfig from './get-generated-files/get-editor-ignore';
import getEslintConfig from './get-generated-files/get-eslint';
import getGithubSyncMetadataConfig from './get-generated-files/get-github-sync-metadata';
import getGithubWorkflowConfig from './get-generated-files/get-github-workflow';
import getGitignoreConfig from './get-generated-files/get-gitignore';
import getGitpodDockerfile from './get-generated-files/get-gitpod-dockerfile';
import getLicenseString from './get-generated-files/get-license-string';
import getLintStaged from './get-generated-files/get-lint-staged';
import getPackageConfig from './get-generated-files/get-package-config';
import getReadmeString from './get-generated-files/get-readme-string';
import getReleaseConfig from './get-generated-files/get-release';
import getRenovateConfig from './get-generated-files/get-renovate';
import getTypescriptConfig from './get-generated-files/get-typescript';
import getVscodeConfig from './get-generated-files/get-vscode';
import githubCodespacesConfig from './get-generated-files/github-codespaces';
import getGitInfo from './get-git-info';
import { PartialTestOptions } from './commands/partial-test-options';

type HandlerWithBase<TConfig extends Config = Config> = (
  this: Base<TConfig>,
  ...args: unknown[]
) => unknown;

type PartialCommandObjectInObjectWithBase<TConfig extends Config = Config> =
  Omit<PartialCommandObjectInObject, 'handler'> & {
    handler: (this: Base<TConfig>, ...args: unknown[]) => unknown;
  };

type PartialCommandInObjectWithBase<TConfig extends Config = Config> =
  | PartialCommandObjectInObjectWithBase<TConfig>
  | HandlerWithBase<TConfig>;

// Extract parameter types from command handlers
type ExtractHandlerParams<T> = T extends {
  handler: (this: any, ...args: infer P) => any;
}
  ? P
  : T extends (this: any, ...args: infer P) => any
  ? P
  : unknown[];

// Get command names from the commands object
type CommandNames<TConfig extends Config> = keyof TConfig['commands'];

// Get parameters for a specific command
type CommandParams<
  TConfig extends Config,
  TCommandName extends CommandNames<TConfig>
> = ExtractHandlerParams<TConfig['commands'][TCommandName]>;

type Config = {
  name?: string;
  global: boolean;
  allowedMatches: string[];
  commands: Record<string, PartialCommandObjectInObjectWithBase>;
  depcheckConfig: Omit<DepcheckOptions, 'package'>;
  deployAssets: Array<{ label: string; path: string }>;
  deployEnv: Record<string, string>;
  deployPlugins: string[];
  editorIgnore: string[];
  fetchGitHistory: boolean;
  git?: GitHost;
  gitignore: string[];
  hasTypescriptConfigRootAlias: boolean;
  lintStagedConfig: LintStagedConfig;
  lint: (options?: PartialCommandOptions) => unknown;
  typecheck: (options?: PartialCommandOptions) => unknown;
  macos: boolean;
  minNodeVersion: number;
  nodeVersion: number;
  preDeploySteps: string[];
  prepare: (options?: PartialCommandOptions) => unknown;
  readmeInstallString: string;
  seeAlso: Array<{ description: string; repository: string }>;
  supportedNodeVersions: number[];
  syncKeywords: boolean;
  typescriptConfig: TsConfigJson;
  windows: boolean;
  testInContainer: boolean;
  eslintConfig: string;
  useJobMatrix: boolean;
  packageConfig: PackageJson;
  renovateConfig: RenovateConfig;
  isLockFileFixCommitType: boolean;
  doppler: boolean;
  npmPublish: boolean;
  codecovGraphToken: string | null;
};

type PartialConfigObject<TConfig extends Config = Config> = Omit<
  Partial<TConfig>,
  'commands'
> & { commands?: Record<string, PartialCommandInObjectWithBase<TConfig>> };

type PartialConfigOrFunction<TConfig extends Config = Config> =
  | PartialConfigObject<TConfig>
  | ((this: Base<TConfig>, config: TConfig) => PartialConfigObject<TConfig>);

type PartialConfig<TConfig extends Config = Config> =
  | PartialConfigOrFunction<TConfig>
  | string
  | null;

export const defineBaseConfig = <T>(configInput: T): T => configInput;

const mergeConfigs = createDefu(<T, K extends keyof T>(obj: T, key: K, value: T[K]) => {
  if (key === 'supportedNodeVersions') {
    obj[key] = value;
    return true;
  }

  return false;
});

class Base<TConfig extends Config = Config> {
  config: TConfig;
  packageConfig: PackageJson;
  cwd: string;
  generatedFiles;
  githubCodespacesConfig = githubCodespacesConfig;

  commit(this: Base, options?: PartialCommandOptions & { allowEmpty?: boolean }) {
    return commit.call(this, options);
  }

  lint(options?: PartialCommandOptions) {
    return lint.call(this, options);
  }

  lintPackagejson() {
    return lintPackagejson.call(this);
  }

  typecheck(options?: PartialCommandOptions) {
    return typecheck.call(this, options);
  }

  verify(options?: PartialCommandOptions & { patterns?: string[] }) {
    return verify.call(this, options);
  }

  prepare(options?: PartialCommandOptions) {
    return prepare.call(this, options);
  }

  test(options?: PartialTestOptions) {
    return test.call(this, options);
  }

  testRaw(options?: PartialTestOptions) {
    return testRaw.call(this, options);
  }

  testDocker(options?: PartialTestOptions) {
    return testDocker.call(this, options);
  }

  getPackageConfig() {
    return getPackageConfig.call(this);
  }

  getGeneratedFiles() {
    return getGeneratedFiles.call(this);
  }

  checkUnknownFiles() {
    return checkUnknownFiles.call(this);
  }

  depcheck() {
    return depcheckMethod.call(this);
  }

  getEditorIgnoreConfig() {
    return getEditorIgnoreConfig.call(this);
  }

  getEslintConfig() {
    return getEslintConfig.call(this);
  }

  getGithubSyncMetadataConfig() {
    return getGithubSyncMetadataConfig.call(this);
  }

  getGithubWorkflowConfig() {
    return getGithubWorkflowConfig.call(this);
  }

  getGitignoreConfig() {
    return getGitignoreConfig.call(this);
  }

  getGitpodDockerfile() {
    return getGitpodDockerfile.call(this);
  }

  getLicenseString() {
    return getLicenseString.call(this);
  }

  getReadmeString() {
    return getReadmeString.call(this);
  }

  getReleaseConfig() {
    return getReleaseConfig.call(this);
  }

  getRenovateConfig() {
    return getRenovateConfig.call(this);
  }

  getVscodeConfig() {
    return getVscodeConfig.call(this);
  }

  getTypescriptConfig() {
    return getTypescriptConfig.call(this);
  }

  getLintStaged() {
    return getLintStaged.call(this);
  }

  constructor(configInput: PartialConfig<TConfig> = null, { cwd = '.' } = {}) {
    this.cwd = cwd;
    const jitiInstance = createJiti(pathLib.resolve(this.cwd));

    const config = (() => {
      if (configInput === null) {
        return { name: packageName`@dword-design/base-config-node` };
      } else if (typeof configInput === 'string') {
        return { name: configInput };
      } else if (typeof configInput === 'function') {
        return configInput.call(this);
      }

      return configInput;
    })();

    if (config.name) {
      config.name = pluginNameToPackageName(config.name, 'base-config');
    }

    this.packageConfig = fs.existsSync(pathLib.join(this.cwd, 'package.json'))
      ? fs.readJsonSync(pathLib.join(this.cwd, 'package.json'))
      : {};

    const defaultConfig = {
      allowedMatches: [],
      commands: {},
      depcheckConfig: {
        detectors: [
          depcheck.detector.exportDeclaration,
          depcheck.detector.importDeclaration,
          depcheck.detector.requireCallExpression,
          depcheck.detector.requireResolveCallExpression,
          depcheckDetectorExeca,
          depcheckDetectorPackageName,
          depcheckDetectorBinName,
        ],
        ignorePath: '.gitignore',
        parsers: { '**/*.ts': depcheck.parser.typescript },
        specials: [getDepcheckSpecialBase(config.name), depcheck.special.bin],
      },
      deployAssets: [],
      deployEnv: {},
      deployPlugins: [],
      doppler: false,
      editorIgnore: [],
      eslintConfig: '',
      fetchGitHistory: false,
      git: getGitInfo({ cwd: this.cwd }),
      gitignore: [],
      global: false,
      hasTypescriptConfigRootAlias: true,
      isLockFileFixCommitType: false,
      lint: () => {},
      lintStagedConfig: {},
      macos: true,
      minNodeVersion: null,
      codecovGraphToken: null,
      nodeVersion: 20,
      packageConfig: {},
      preDeploySteps: [],
      prepare: identity,
      readmeInstallString: endent`
        ## Install

        \`\`\`bash
        # npm
        $ npm install ${config.global ? '-g ' : ''}${this.packageConfig.name}

        # Yarn
        $ yarn ${config.global ? 'global ' : ''}add ${this.packageConfig.name}
        \`\`\`
      `,
      renovateConfig: {},
      seeAlso: [],
      supportedNodeVersions: [20, 22],
      syncKeywords: true,
      testInContainer: false,
      typecheck: () => {},
      typescriptConfig: {},
      useJobMatrix: true,
      windows: true,
    };

    const inheritedConfigPath =
      config.name && config.name === this.packageConfig.name
        ? pathLib.resolve(this.cwd, 'src', 'index.ts')
        : config.name;

    let inheritedConfig:
      | PartialConfigOrFunction<TConfig>
      | { default: PartialConfigOrFunction<TConfig> } = inheritedConfigPath
      ? jitiInstance(inheritedConfigPath)
      : undefined;

    if (inheritedConfig && 'default' in inheritedConfig) {
      inheritedConfig = inheritedConfig.default;
    }

    if (typeof inheritedConfig === 'function') {
      inheritedConfig = inheritedConfig.call(
        this,
        mergeConfigs(defaultConfig, config),
      );
    }

    this.config = mergeConfigs(config, inheritedConfig, defaultConfig);

    this.config = {
      ...this.config,
      commands: mapValues(this.config.commands, command =>
        typeof command === 'function' ? { handler: command } : command,
      ),
    };

    this.packageConfig = this.getPackageConfig();
    this.generatedFiles = this.getGeneratedFiles();
  }

  run<TCommandName extends keyof TConfig['commands'] & string>(
    name: TCommandName,
    ...args: CommandParams<TConfig, TCommandName>
  ) {
    return this.config.commands[name].handler.call(this, ...args);
  }
}

export { default as loadConfig } from './load-config';

export { default as loadConfigSync } from './load-config-sync';

export { Base };

export type { Config, PartialConfig };

export { type PartialCommandOptions } from './commands/partial-command-options';
