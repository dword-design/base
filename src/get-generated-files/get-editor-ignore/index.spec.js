import tester from '@dword-design/tester'

import { Base } from '@/src'

export default tester(
  {
    valid: ['foo'],
  },
  [
    {
      transform: test =>
        function () {
          const base = new Base({ editorIgnore: test })
          expect(base.getEditorIgnore()).toMatchSnapshot(this)
        },
    },
  ]
)
