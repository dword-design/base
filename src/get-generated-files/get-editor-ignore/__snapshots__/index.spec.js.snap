// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index valid 1`] = `
Array [
  ".babelrc.json",
  ".commitlintrc.json",
  ".cz.json",
  ".editorconfig",
  ".eslintrc.json",
  ".gitattributes",
  ".github",
  ".gitignore",
  ".gitpod.Dockerfile",
  ".gitpod.yml",
  ".husky",
  ".nyc_output",
  ".releaserc.json",
  ".renovaterc.json",
  "CHANGELOG.md",
  "LICENSE.md",
  "coverage",
  "foo",
  "node_modules",
  "yarn.lock",
]
`;
